maggma/examples/builders.py

```diff
@@ -0,0 +1,274 @@
+"""
+Example builders for testing and general use.
+"""
+import traceback
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+
+from pydash import identity
+
+from maggma.builder import Builder
+from maggma.utils import confirm_field_index, total_size
+
+
+def source_keys_updated(source, target):
+    """
+    Utility for incremental building. Gets a list of source.key values.
+
+    Get key values for source documents that have been updated with respect to
+    corresponding target documents.
+
+    Called should ensure [(lu_field, -1),(key, 1)] compound index on both source
+    and target.
+
+    """
+    keys_updated = set()  # Handle non-unique keys, e.g. for GroupBuilder.
+    cursor_source = source.query(
+        properties=[source.key, source.lu_field],
+        sort=[(source.lu_field, -1), (source.key, 1)])
+    cursor_target = target.query(
+        properties=[target.key, target.lu_field],
+        sort=[(target.lu_field, -1), (target.key, 1)])
+    tdoc = next(cursor_target, None)
+    for sdoc in cursor_source:
+        if tdoc is None:
+            keys_updated.add(sdoc[source.key])
+            continue
+
+        if tdoc[target.key] == sdoc[source.key]:
+            if tdoc[target.lu_field] < source.lu_func[0](sdoc[source.lu_field]):
+                keys_updated.add(sdoc[source.key])
+            tdoc = next(cursor_target, None)
+        else:
+            keys_updated.add(sdoc[source.key])
+    return list(keys_updated)
+
+
+def get_criteria(source, target, query=None, incremental=True, logger=None):
+    """Return criteria to pass to `source.query` to get items."""
+    index_checks = [confirm_field_index(target, target.key)]
+    if incremental:
+        # Ensure [(lu_field, -1), (key, 1)] index on both source and target
+        for store in (source, target):
+            info = store.collection.index_information().values()
+            index_checks.append(
+                any(spec == [(store.lu_field, -1), (store.key, 1)]
+                    for spec in (index['key'] for index in info)))
+    if not all(index_checks):
+        index_warning = (
+            "Missing one or more important indices on stores. "
+            "Performance for large stores may be severely degraded. "
+            "Ensure indices on target.key and "
+            "[(store.lu_field, -1), (store.key, 1)] "
+            "for each of source and target."
+        )
+        logger.warning(index_warning)
```

Comment:
Won't this break when logger is left as the default?

Reply:
break? the WARNING log level is higher than info, so a default log level of INFO will pass this. Is the default level higher, e.g. ERROR? I thought INFO was the default.

Reply:
Right, but the default value for logger is None, and I don't see it being checked or assigned anywhere.

Reply:
From [maggma.builder.Builder](https://github.com/materialsproject/maggma/blob/c4a4ae5db28d6f3eb9dcb0d745873144fa1bf7c0/maggma/builder.py#L35):
```
        self.logger = logging.getLogger(type(self).__name__)
        self.logger.addHandler(logging.NullHandler())
```
The logger is defined and won't "break", but there might be no handler defined, in which case there might not be any output? But calls to logger methods are fine.